[ML] Data Frames - KQL bar for querying source data in creation wizard

alvarezmelissa87 opened the pull request:
## Summary

Use KQL bar for source data querying in Data Frames creation wizard.

Keeping the old structure in place for future use of toggling between using Lucene and KQL. Default is set to KQL for now.

NOTE:
This PR also includes small change to the `KqlFilterBar` component to make it more generic. It removes placeholder text specific to use in explorer and renames the created query to `filterQuery` instead of the more specific `influencersFilterQuery`. Explorer has been updated accordingly.

![image](https://user-images.githubusercontent.com/6446462/61723623-71c5af00-ad3a-11e9-8fd1-4dfedc5e3aca.png)

Autocomplete for fields:

![image](https://user-images.githubusercontent.com/6446462/61723656-8013cb00-ad3a-11e9-9f69-a4c76b26757d.png)

![image](https://user-images.githubusercontent.com/6446462/61723678-899d3300-ad3a-11e9-83b4-6f4ddcef4606.png)

![image](https://user-images.githubusercontent.com/6446462/61723693-93269b00-ad3a-11e9-9998-ea3f4be49b87.png)

Allows more complex queries specific to numerical values:

![image](https://user-images.githubusercontent.com/6446462/61723727-9e79c680-ad3a-11e9-98f2-19818142f76d.png)


### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
~~- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~~
~~- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [x] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

~~- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~
~~- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~

elasticmachine commented:
Pinging @elastic/ml-ui

peteharverson commented on x-pack/legacy/plugins/ml/public/components/kql_filter_bar/filter_bar/filter_bar.js:
```diff
@@ -169,7 +169,7 @@ export class FilterBar extends Component {
               'xpack.ml.explorer.kueryBar.filterPlaceholder',
```
Does this need to be a translation any more? Aren't you just passing the `placeholder` prop straight through now?

alvarezmelissa87 replied:
It does not - removed in c6a1e15c98cb28c5aedb9f6530a0ded4e8cd73b9

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8914/)

peteharverson commented on x-pack/legacy/plugins/ml/public/data_frame/pages/data_frame_new_pivot/components/step_define/step_define_form.tsx:
```diff
@@ -441,19 +451,17 @@ export const StepDefineForm: SFC<Props> = React.memo(({ overrides = {}, onChange
+                      'xpack.ml.dataframe.stepDefineForm.queryPlaceholder',
+                      {
+                        defaultMessage: 'e.g. {example}',
+                        values: { example: 'method:GET -is:active' },
```
The `-is:active` part of the placeholder is not valid KQL is it? I'd remove that part of the placeholder and either just leave `method:GET` or make it something like `method : "GET" or status : "404"`

alvarezmelissa87 replied:
Good call - updated in 0f1f18012dc0bd4505fa88b6d113eb080915314a

peteharverson commented:
Testing this out, I am seeing issues when moving forward and back in the wizard, where the query possibly isn't being retained properly.

For example, using cloudwatch data, I add a query of `region : "sa-east-1" or region : "ap-northwest-1" `, then add a pivot of `terms(instance)` and aggregation of `avg(CPUUtilization)`. Works as expected:

![image](https://user-images.githubusercontent.com/7405507/61728618-7d7e9900-ad6d-11e9-8389-682d2258f890.png)

Then I hit 'Next', but lose the Pivot Preview:
![image](https://user-images.githubusercontent.com/7405507/61728671-98510d80-ad6d-11e9-923d-1fe62a7ba379.png)

Then hit 'Previous' and now the source index preview isn't displaying either:
![image](https://user-images.githubusercontent.com/7405507/61728738-b0289180-ad6d-11e9-89a4-6b73e2ef2d74.png)

alvarezmelissa87 commented:
Good catch @peteharverson - the search query was initially set to `search` and I updated to `searchQuery` - I've fixed the issue by ensuring each step is pulling the correct query - `searchQuery` (including the query body built for the create request). 👍

alvarezmelissa87 commented:
This has been updated 👌 cc @peteharverson, @walterra, @jgowdyelastic

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8978/)

walterra commented on x-pack/legacy/plugins/ml/public/explorer/explorer.js:
```diff
@@ -789,7 +789,8 @@ export const Explorer = injectI18n(injectObservablesAsProps(
           if (stateUpdate.influencers[influencerName][0] && stateUpdate.influencers[influencerName][0].influencerFieldValue) {
-            stateUpdate.filterPlaceHolder = `${influencerName} : ${stateUpdate.influencers[influencerName][0].influencerFieldValue}`;
+            stateUpdate.filterPlaceHolder =
+            `Filter by influencer fields... (E.g. ${influencerName} : ${stateUpdate.influencers[influencerName][0].influencerFieldValue})`;
```
I suspect this now needs to be translated? Can you maybe reuse the translation id `xpack.ml.explorer.kueryBar.filterPlaceholder` if it's the same message and was just moved to another place?

peteharverson commented on x-pack/legacy/plugins/ml/public/data_frame/pages/data_frame_new_pivot/components/step_define/step_define_form.tsx:
```diff
@@ -441,19 +458,17 @@ export const StepDefineForm: SFC<Props> = React.memo(({ overrides = {}, onChange
                     defaultMessage: 'Use a query string to filter the source data (optional).',
```
I think we should drop `string` from this label seeing as it is now using KQL, to leave `Use a query to filter the source data (optional).`

jgowdyelastic commented on x-pack/legacy/plugins/ml/public/data_frame/pages/data_frame_new_pivot/components/step_define/step_define_form.tsx:
```diff
@@ -35,6 +35,8 @@ import { AggListForm } from '../aggregation_list';
 import { SourceIndexPreview } from '../source_index_preview';
 import { PivotPreview } from './pivot_preview';
+// @ts-ignore: could not find declaration file for module
+import { KqlFilterBar } from '../../../../../components/kql_filter_bar';
```
is it worth adding a definition file for this rather than suppressing the error. Looking at the class it'd be something like:
```
import { FC } from 'react';
declare const KqlFilterBar: FC<{
  indexPattern: any;
  initialValue?: string;
  onSubmit: (kqlQueryValues: any) => void;
  placeholder?: string;
  valueExternal?: string;
}>;

```
`indexPattern` and `onSubmit` could have better types if it's easy to do.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9186/)
